src/main/java/com/microsoft/azure/vmagent/AzureVMManagementServiceDelegate.java

```diff
@@ -1692,36 +1694,54 @@ public boolean isVMAliveOrHealthy(AzureVMAgent agent) throws AzureCloudException
      * @return Total VM count
      */
     public int getVirtualMachineCount(String cloudName, String resourceGroupName) {
+    	final Map<String, Integer> countsByTemplate = getVirtualMachineCountsByTemplate(cloudName, resourceGroupName);
+        int count = 0;
+        for (Integer countForTemplate : countsByTemplate.values() ) {
+        	count += countForTemplate.intValue();
+        }
+        return count;
+    }
+
+    /**
+     * Retrieves count of virtual machine in a azure subscription indexed by template name.
+     *
+     * @return {@link Map} of {@link String} template name to {@link Integer} count of VMs from that template.
+     * @throws ManagementException if we get an Azure error that isn't 404.
+     */
+    @Restricted(NoExternalUse.class)
+    Map<String, Integer> getVirtualMachineCountsByTemplate(final String cloudName, final String resourceGroupName) {
+    	final Map<String, Integer> result = new TreeMap<>();
         try {
             final PagedIterable<VirtualMachine> vms = azureClient.virtualMachines().listByResourceGroup(resourceGroupName);
-            int count = 0;
             final AzureUtil.DeploymentTag deployTag = new AzureUtil.DeploymentTag();
-            for (VirtualMachine vm : vms) {
+            for (final VirtualMachine vm : vms) {
                 final Map<String, String> tags = vm.tags();
-                if (tags.containsKey(Constants.AZURE_RESOURCES_TAG_NAME)
-                    && deployTag.isFromSameInstance(
-                    new AzureUtil.DeploymentTag(tags.get(Constants.AZURE_RESOURCES_TAG_NAME)))) {
-                    if (tags.containsKey(Constants.AZURE_CLOUD_TAG_NAME)) {
-                        if (tags.get(Constants.AZURE_CLOUD_TAG_NAME).equals(cloudName)) {
-                            count++;
-                        }
-                    } else {
-                        // keep backwards compatibility, omitting the resource created before updates
-                        // until all the resources has cloud tag
-                        count++;
-                    }
+                final String resourcesTag = tags.containsKey(Constants.AZURE_RESOURCES_TAG_NAME) ? tags.get(Constants.AZURE_RESOURCES_TAG_NAME) : null;
+                final String cloudTag = tags.containsKey(Constants.AZURE_CLOUD_TAG_NAME) ? tags.get(Constants.AZURE_CLOUD_TAG_NAME) : null;
+                final String templateTag = tags.containsKey(Constants.AZURE_TEMPLATE_TAG_NAME) ? tags.get(Constants.AZURE_TEMPLATE_TAG_NAME) : null;
+                if (resourcesTag == null || cloudTag == null || !deployTag.isFromSameInstance(new AzureUtil.DeploymentTag(resourcesTag))) {
```

Comment:
The previous code contained "kept for backwards compatibilty" stuff that I've removed.
The old code wasn't sure that everyone would have `cloudTag` set or not ... but that was in 2017, so I think we're safe now.